### DISCUSSION TITLE
Celery integration

### DISCUSSION THREAD

**Discussion by underyx** [hynek/svcs#69] [Ideas]
It would be cool to be able to request svcs for the lifetime of a Celery task. The idea of having this was actually what really drove the point home for me as to why svcs is an awesome idea; I have a Flask + Celery codebase and it would be amazing to be able to use utility functions that grab svcs in the correct way for themselves depending on which context they're called in.

**Comment by hynek** [comment]
Yes, providing a unified dependency abstraction is indeed the goal. The problem is, I know nothing about Celery. This is something a domain expert will have to at least design.

**Comment by atbigelow** [comment]
I took a stab at writing this! I am certainly not an expert with Celery internals, though.

https://github.com/atbigelow/svcs/tree/celery-integration

How to get proper hooks into Celery for holding `Registry` and `Container` state has been frustrating. I eventually created signal handlers that create a `Container` on the `celery.Task.request` instance. These signals will fire regardless if you've `init`'d a Celery instance, though. The handlers I created peer into the Task and app to check for a `Registry`.

I am going to keep working on this to see if this really is the best way to hook in. But do plan on making a PR in the near future.

**Reply by hynek** [reply]
Thanks for working on it! As I mentioned in another similar thread: I do not plan adding more integrations into svcs proper, because I can't deal with the maintenance burden – especially of integrations I don't use myself and have no domain knowledge about.

I'm happy to assist you to wrap the resulting code into a package like https://github.com/django-svcs/django-svcs or https://github.com/guacs/litestar-svcs, though!

**Comment by atbigelow** [comment]
Sounds great! That's the recommended route, so I'll do that.

Maybe a side discussion to have about how best to surface these not-included integrations for visibility and ease of discovery.

**Reply by hynek** [reply]
There is https://github.com/hynek/svcs/wiki/Third%E2%80%90party-Extensions which is linked from the readme and from the docs side bar but I’m very open to suggestions to make people more aware of it. A lively eco system would be awesome.

**Reply by atbigelow** [reply]
Haha, doh! Tunnel vision I guess 😂. If I can think of anything, I'll say something. Thanks again!
